pycloudlib/openstack/instance.py

```diff
@@ -0,0 +1,151 @@
+"""Openstack instance type."""
+import time
+from itertools import chain
+
+import openstack
+from openstack.exceptions import (
+    BadRequestException,
+    ConflictException,
+    ResourceNotFound,
+)
+
+from pycloudlib.instance import BaseInstance
+
+
+class OpenstackInstance(BaseInstance):
+    """Openstack instance object."""
+
+    _type = 'openstack'
+
+    def __init__(self, key_pair, instance_id, connection=None):
+        """Set up the instance.
+
+        Args:
+            key_pair: A KeyPair for SSH interactions
+            instance_id: The instance id representing the cloud instance
+            connection: The connection used to create this instance.
+                If None, connection will be created.
+        """
+        super().__init__(key_pair)
+
+        if not connection:
+            connection = openstack.connect()
+        self.conn = connection
+
+        self.server = self.conn.compute.get_server(instance_id)
+
+        self.delete_floating_ip = False
+        self.floating_ip = self._get_existing_floating_ip()
+        if self.floating_ip is None:
+            self._create_and_attach_floating_id()
```

Comment:
A nit, but `self.floating_ip = self._create_and_attach_floating_id()` feels a little easier to follow; it would give `_get_...` and `_create_...` a (more) similar API and means that `self.floating_ip` is only set in one place.